### PR TITLE
feat: Enable Spark SQL tests for Spark 3.5.1

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/589

## What changes are included in this PR?

- Add diff file for 3.5.1
- Enable Spark SQL tests for 3.5.1 in CI
- Add documentation on how I did this

### CODE REVIEW

File: docs/source/contributor-guide/spark-sql-tests.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Running Spark SQL Tests
+
+Running Apache Spark's SQL tests with Comet enabled is a good way to ensure that Comet produces the same 
+results as that version of Spark. To do this, we need to  make some changes to the Apache Spark source code so that 
+it enables Comet.
```
Comment (kazuyukitanimura) [inline docs/source/contributor-guide/spark-sql-tests.md:24]: `we need to make some changes`
Technically we do not need to make changes in Spark to enable Comet, but for testing purpose, we implemented some convenient features.

File: docs/source/contributor-guide/spark-sql-tests.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+For example:
+
+```shell
+wiggle --replace ./sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala ./sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala.rej
+```
+
+## Generating The Diff File
+
+```shell
+git diff v3.5.1 58c7ce4407d2c4685a1feaf3e60cefac32de0d39 > ../datafusion-comet/dev/diffs/3.5.1.diff
```
Comment (kazuyukitanimura) [inline docs/source/contributor-guide/spark-sql-tests.md:103]: Can be simply `git diff v3.5.1 > ..`

**Comment by andygrove** [comment]
@kazuyukitanimura Spark SQL tests are now passing in CI, but I had to ignore quite a few. I will file a follow on issue for resolving those.

**Review by kazuyukitanimura — COMMENTED** [review]
Looks good.
What would be the best way to find ignored tests that are specific to 3.5?
For example, https://github.com/apache/datafusion-comet/blob/main/dev/diffs/4.0.0-preview1.diff#L445 
Spark 4.0.0-preview1.diff has #551 mentioned for all additionally ignored tests compared 3.4.3.diff

**Comment by andygrove** [comment]
> Looks good. What would be the best way to find ignored tests that are specific to 3.5? For example, https://github.com/apache/datafusion-comet/blob/main/dev/diffs/4.0.0-preview1.diff#L445 Spark 4.0.0-preview1.diff has #551 mentioned for all additionally ignored tests compared 3.4.3.diff

Good point. I will go ahead and update the diff with links for all newly ignored tests compared to 3.4.3.

**Review by kazuyukitanimura — APPROVED** [review]
LGTM

**Comment by andygrove** [comment]
Thanks for the review @kazuyukitanimura

**Pull request merged**